src/ucp/stream/stream_recv.c

```diff
@@ -226,48 +251,58 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_nb,
                  size_t *length, unsigned flags)
 {
     ucp_ep_ext_stream_t *ep_stream = ep->ext.stream;
-    ucp_request_t       tmp_req;
+    size_t              dt_length;
     ucp_request_t       *req;
     ucs_status_ptr_t    rdesc;
     ucs_status_t        status;
+    ucp_dt_state_t      dt_state;
 
     if (UCP_DT_IS_GENERIC(datatype)) {
         ucs_error("ucp_stream_recv_nb doesn't support generic datatype");
         return UCS_STATUS_PTR(UCS_ERR_NOT_IMPLEMENTED);
     }
 
+    dt_length = ucp_dt_length(datatype, count, buffer, &dt_state);
+
     UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
 
-    if (ucs_unlikely(!(ep_stream->flags & UCP_EP_STREAM_FLAG_HAS_DATA))) {
-        /* No data, enqueue request and out */
-        req = ucp_request_get(ep->worker);
-        if (ucs_unlikely(req == NULL)) {
-            req = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
-            goto out;
-        }
-        ucp_stream_recv_request_init(req, buffer, count, datatype, cb);
-        ucs_queue_push(&ep_stream->match_q, &req->recv.queue);
-        goto ptr_out;
+    if (ucs_likely(ucp_stream_recv_nb_is_inplace(ep_stream, dt_length))) {
+        ucp_dt_recv_state_init(&dt_state, buffer, datatype, count);
+        status = ucp_stream_process_rdesc_inplace(ucp_stream_rdesc_get(ep_stream),
+                                                  datatype, buffer, dt_length,
+                                                  ep_stream, &dt_state);
+        req = UCS_STATUS_PTR(status);
+        *length = dt_length;
+        goto out;
+    }
+
+    req = ucp_request_get(ep->worker);
+    if (ucs_unlikely(req == NULL)) {
+        req = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+        goto out;
     }
 
-    /* We can complete request in place, use tmp request on stack */
-    req = &tmp_req;
-    ucp_stream_recv_request_init(req, buffer, count, datatype, cb);
+    ucp_stream_recv_request_init(req, buffer, count, dt_length, datatype, cb);
 
     /* OK, lets obtain all arrived data which matches the recv size */
-    do {
-        rdesc  = ucp_stream_recv_data_nb_internal(ep_stream, 0);
+    while ((req->recv.state.offset < req->recv.length) &&
+           (ep_stream->flags & UCP_EP_STREAM_FLAG_HAS_DATA)) {
+
+        rdesc  = ucp_stream_rdesc_get(ep_stream);
         status = ucp_stream_process_rdesc(rdesc, ep_stream, req);
         if (ucs_unlikely(status != UCS_OK)) {
-            req = UCS_STATUS_PTR(status);
-            goto out;
+            goto bad_req;
         }
-    } while ((req->recv.state.offset < req->recv.length) &&
-             (ep_stream->flags & UCP_EP_STREAM_FLAG_HAS_DATA));
+    }
 
-    ucs_assert(req->recv.state.offset >  0);
     ucs_assert(req->recv.state.offset <= req->recv.length);
+
+    if (!ucp_request_can_complete_stream_recv(req)) {
```

Comment:
i suggest to simplify the exit/error flow as follows:
```

	while (... ) {
		if (unlikely(status != UCS_OK)) {
			goto out_status;
		}
	}
	
	if (!ucp_request_can_complete_stream_recv(req)) {
		ucs_assert(...);
		ucs_queue_push(...);
		goto out;
	}

	*length = ...
	status = UCS_OK;

out_status:
	ucp_request_put(req);
	req = UCS_STATUS_PTR(status);
out:
	UCP_THREAD_CS_EXIT (...)
	return req;
```

Reply:
We have path without request, so seems like third label is needed.

Reply:
do you mean like this?
```
out_put_request:
	ucp_request_put(req);
out_status:
	req = UCS_STATUS_PTR(status);
out:
	UCP_THREAD_CS_EXIT (...)
	return req;
```